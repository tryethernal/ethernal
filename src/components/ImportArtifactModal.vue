<template>
<v-dialog v-model="dialog" max-width="600">
    <v-card>
        <v-card-title class="headline">Update Contract Metadata</v-card-title>

        <v-card-text>
            <v-alert type="success" v-if="successMessage" v-html="successMessage"></v-alert>
            <v-alert type="error" v-if="errorMessage" v-html="errorMessage"></v-alert>
            <div class="mb-3">
                Edit the name and the ABI of this contract to be able to interact with its read/write methods, decode functions calls & events emitted.<br>
                Only the formatting of the ABI is checked, it's up to you to make sure it is valid for the current address. If it's not, calls will fail.
            </div>
            <div>
                <v-text-field id="contractName" class="mb-3" primary hide-details="auto" outlined dense v-model="name" label="Contract Name"></v-text-field>
                <v-textarea id="contractAbi" primary hide-details="auto" outlined dense v-model="abi" label="ABI"></v-textarea>
                <v-divider class="my-3"></v-divider>
                Or upload a json file containing a <b>contractName</b> field and/or an <b>abi</b> field (a Truffle or Hardhat build file for example):<br>
                <input type="file" id="fileUploader" ref="fileUploader" accept="application/json" v-on:change="handleFileUpload" />
            </div>
        </v-card-text>

        <v-card-actions>
            <v-spacer></v-spacer>
            <v-btn color="primary" text @click.stop="close()">Close</v-btn>
            <v-btn id="updateContract" color="primary" :loading="loading" text @click.stop="update()">Update</v-btn>
        </v-card-actions>
    </v-card>
</v-dialog>
</template>
<script>
const ethers = require('ethers');
import { mapGetters } from 'vuex';

export default {
    name: 'ImportArtifactModal',
    data: () => ({
        dialog: false,
        resolve: null,
        reject: null,
        successMessage: null,
        errorMessage: null,
        loading: false,
        abi: null,
        name: null,
        address: null,
        updated: false
    }),
    methods: {
        open: function(options) {
            this.dialog = true;
            this.address = options.address;
            this.name = options.name;
            this.abi = options.abi && options.abi.trim().length ? options.abi : '';
            return new Promise((resolve, reject) => {
                this.resolve = resolve;
                this.reject = reject;
            });
        },
        onFileLoaded: function(data) {
            const parsedArtifact = JSON.parse(data.target.result);
            if (parsedArtifact.abi) {
                this.abi = JSON.stringify(parsedArtifact.abi);
            }
            if (parsedArtifact.contractName || parsedArtifact.name) {
                this.name = parsedArtifact.contractName || parsedArtifact.name;
            }
            if (!parsedArtifact.abi && !parsedArtifact.contractName) {
                this.errorMessage = `No abi or contractName fields found. Please upload a valid file.`;
            }
        },
        handleFileUpload: function(ev) {
            const file = ev.target.files[0];
            if (!file)
                return;

            const fileReader = new FileReader();
            fileReader.onload = this.onFileLoaded;
            fileReader.readAsText(file);
        },
        update: function() {
            this.loading = true;
            this.successMessage = null;
            this.errorMessage = null;
            let parsedAbi;

            try {
                parsedAbi = this.abi && this.abi.trim().length ?
                    JSON.parse(this.abi) :
                    '';
            } catch(_) {
                this.loading = false;
                return this.errorMessage = 'Invalid JSON format';
            }

            if (typeof parsedAbi == 'object') {
                try {
                    !!this.abi && new ethers.utils.Interface(parsedAbi);
                } catch(_) {
                    this.loading = false;
                    return this.errorMessage = 'Invalid ABI';
                }
            }

            this.server.syncContractData(this.address, this.name, parsedAbi)
                .then(() => {
                    this.updated = true;
                    this.successMessage = 'Metadata updated';
                })
                .catch((error) => this.errorMessage = error)
                .finally(() => this.loading = false);
        },
        close: function() {
            this.resolve(this.updated);
            this.reset();
        },
        reset: function() {
            this.dialog = false;
            this.resolve = null;
            this.reject = null;
            this.name = null;
            this.abi = null;
            this.address = null;
            this.successMessage = null;
            this.errorMessage = null;
        }
    },
    computed: {
        ...mapGetters([
            'currentWorkspace'
        ])
    }
}
</script>
