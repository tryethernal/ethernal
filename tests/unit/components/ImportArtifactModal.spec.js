import MockHelper from '../MockHelper';

import ImportArtifactModal from '@/components/ImportArtifactModal.vue';
import DSProxyFactoryContract from '../fixtures/DSProxyFactoryContract.json';

describe('ImportArtifactModal.vue', () => {
    let helper;

    beforeEach(async () => {
        helper = new MockHelper();
    });

    it('Should let the user edit the name and ABI when parameters are valid', async () => {
        const wrapper = helper.mountFn(ImportArtifactModal);
        await wrapper.setData({ dialog: true });

        const syncContractDataMock = jest.spyOn(helper.mocks.server, 'syncContractData');

        await wrapper.find('#contractName').setValue(DSProxyFactoryContract.name);
        await wrapper.find('#contractAbi').setValue(JSON.stringify(DSProxyFactoryContract.abi));
        await wrapper.find('#updateContract').trigger('click');

        await new Promise(process.nextTick);

        expect(syncContractDataMock).toHaveBeenCalled();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should warn the user if the ABI is not valid', async () => {
        const wrapper = helper.mountFn(ImportArtifactModal);
        await wrapper.setData({ dialog: true });

        const syncContractDataMock = jest.spyOn(helper.mocks.server, 'syncContractData');

        await wrapper.find('#contractName').setValue(DSProxyFactoryContract.name);
        await wrapper.find('#contractAbi').setValue('{ "a": 1 }');
        await wrapper.find('#updateContract').trigger('click');

        await new Promise(process.nextTick);

        expect(wrapper.vm.errorMessage).toBe('Invalid ABI');
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should warn the user if the ABI is not valid json', async () => {
        const wrapper = helper.mountFn(ImportArtifactModal);
        await wrapper.setData({ dialog: true });

        await wrapper.find('#contractName').setValue(DSProxyFactoryContract.name);
        await wrapper.find('#contractAbi').setValue('{ a: 1 }');
        await wrapper.find('#updateContract').trigger('click');
        await new Promise(process.nextTick);

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should let the user upload a file containing contract metadata', async () => {
        const wrapper = helper.mountFn(ImportArtifactModal);
        await wrapper.setData({ dialog: true });
        const syncContractDataMock = jest.spyOn(helper.mocks.server, 'syncContractData');

        const fileContent = JSON.stringify(DSProxyFactoryContract);
        wrapper.vm.onFileLoaded({ target: { result: fileContent }});

        await wrapper.find('#updateContract').trigger('click');
        await new Promise(process.nextTick);
        
        expect(syncContractDataMock).toHaveBeenCalled();
        expect(wrapper.vm.abi).toBe(JSON.stringify(DSProxyFactoryContract.abi));
        expect(wrapper.vm.name).toBe(DSProxyFactoryContract.name);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should warn the user if the file is invalid', async () => {
        const wrapper = helper.mountFn(ImportArtifactModal);
        await wrapper.setData({ dialog: true });
        const syncContractDataMock = jest.spyOn(helper.mocks.server, 'syncContractData');


        const fileContent = JSON.stringify({ a: 1 });
        wrapper.vm.onFileLoaded({ target: { result: fileContent }});
        await wrapper.find('#updateContract').trigger('click');
        await new Promise(process.nextTick);

        expect(syncContractDataMock).toHaveBeenCalled();
        expect(wrapper.html()).toMatchSnapshot();
    });    
});
