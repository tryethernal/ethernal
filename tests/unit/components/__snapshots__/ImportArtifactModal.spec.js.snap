// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportArtifactModal.vue Should let the user edit the name and ABI when parameters are valid 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Update Contract Metadata</div>
          <div class="v-card__text">
            <div role="alert" class="v-alert v-sheet theme--dark success">Metadata updated</div>
            <!---->
            <div class="mb-3">
              Edit the name and the ABI of this contract to be able to interact with its read/write methods, decode functions calls &amp; events emitted.<br>
              Only the formatting of the ABI is checked, it's up to you to make sure it is valid for the current address. If it's not, calls will fail.
            </div>
            <div>
              <div class="v-input mb-3 v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: -24px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="contractName" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Contract Name</label><input primary="" id="contractName" type="text"></div>
                  </div>
                </div>
              </div>
              <div class="v-input v-textarea v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: -24px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="contractAbi" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">ABI</label><textarea primary="" id="contractAbi" rows="5"></textarea></div>
                  </div>
                </div>
              </div>
              <hr role="separator" aria-orientation="horizontal" class="my-3 v-divider theme--light">
              Or upload a json file containing a <b>contractName</b> field and/or an <b>abi</b> field (a Truffle or Hardhat build file for example):<br> <input type="file" id="fileUploader" accept="application/json">
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text" id="updateContract"><span class="v-btn__content">Update</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`ImportArtifactModal.vue Should let the user upload a file containing contract metadata 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Update Contract Metadata</div>
          <div class="v-card__text">
            <div role="alert" class="v-alert v-sheet theme--dark success">Metadata updated</div>
            <!---->
            <div class="mb-3">
              Edit the name and the ABI of this contract to be able to interact with its read/write methods, decode functions calls &amp; events emitted.<br>
              Only the formatting of the ABI is checked, it's up to you to make sure it is valid for the current address. If it's not, calls will fail.
            </div>
            <div>
              <div class="v-input mb-3 v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: -24px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="contractName" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Contract Name</label><input primary="" id="contractName" type="text"></div>
                  </div>
                </div>
              </div>
              <div class="v-input v-textarea v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: -24px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="contractAbi" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">ABI</label><textarea primary="" id="contractAbi" rows="5"></textarea></div>
                  </div>
                </div>
              </div>
              <hr role="separator" aria-orientation="horizontal" class="my-3 v-divider theme--light">
              Or upload a json file containing a <b>contractName</b> field and/or an <b>abi</b> field (a Truffle or Hardhat build file for example):<br> <input type="file" id="fileUploader" accept="application/json">
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text" id="updateContract"><span class="v-btn__content">Update</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`ImportArtifactModal.vue Should warn the user if the ABI is not valid 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Update Contract Metadata</div>
          <div class="v-card__text">
            <!---->
            <div role="alert" class="v-alert v-sheet theme--dark error">Invalid ABI</div>
            <div class="mb-3">
              Edit the name and the ABI of this contract to be able to interact with its read/write methods, decode functions calls &amp; events emitted.<br>
              Only the formatting of the ABI is checked, it's up to you to make sure it is valid for the current address. If it's not, calls will fail.
            </div>
            <div>
              <div class="v-input mb-3 v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: -24px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="contractName" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Contract Name</label><input primary="" id="contractName" type="text"></div>
                  </div>
                </div>
              </div>
              <div class="v-input v-textarea v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: -24px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="contractAbi" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">ABI</label><textarea primary="" id="contractAbi" rows="5"></textarea></div>
                  </div>
                </div>
              </div>
              <hr role="separator" aria-orientation="horizontal" class="my-3 v-divider theme--light">
              Or upload a json file containing a <b>contractName</b> field and/or an <b>abi</b> field (a Truffle or Hardhat build file for example):<br> <input type="file" id="fileUploader" accept="application/json">
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text" id="updateContract"><span class="v-btn__content">Update</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`ImportArtifactModal.vue Should warn the user if the ABI is not valid json 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Update Contract Metadata</div>
          <div class="v-card__text">
            <!---->
            <div role="alert" class="v-alert v-sheet theme--dark error">Invalid JSON format</div>
            <div class="mb-3">
              Edit the name and the ABI of this contract to be able to interact with its read/write methods, decode functions calls &amp; events emitted.<br>
              Only the formatting of the ABI is checked, it's up to you to make sure it is valid for the current address. If it's not, calls will fail.
            </div>
            <div>
              <div class="v-input mb-3 v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: -24px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="contractName" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Contract Name</label><input primary="" id="contractName" type="text"></div>
                  </div>
                </div>
              </div>
              <div class="v-input v-textarea v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: -24px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="contractAbi" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">ABI</label><textarea primary="" id="contractAbi" rows="5"></textarea></div>
                  </div>
                </div>
              </div>
              <hr role="separator" aria-orientation="horizontal" class="my-3 v-divider theme--light">
              Or upload a json file containing a <b>contractName</b> field and/or an <b>abi</b> field (a Truffle or Hardhat build file for example):<br> <input type="file" id="fileUploader" accept="application/json">
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text" id="updateContract"><span class="v-btn__content">Update</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`ImportArtifactModal.vue Should warn the user if the file is invalid 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Update Contract Metadata</div>
          <div class="v-card__text">
            <div role="alert" class="v-alert v-sheet theme--dark success">Metadata updated</div>
            <!---->
            <div class="mb-3">
              Edit the name and the ABI of this contract to be able to interact with its read/write methods, decode functions calls &amp; events emitted.<br>
              Only the formatting of the ABI is checked, it's up to you to make sure it is valid for the current address. If it's not, calls will fail.
            </div>
            <div>
              <div class="v-input mb-3 v-input--hide-details v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="contractName" class="v-label theme--light" style="left: 0px; position: absolute;">Contract Name</label><input primary="" id="contractName" type="text"></div>
                  </div>
                </div>
              </div>
              <div class="v-input v-textarea v-input--hide-details v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="contractAbi" class="v-label theme--light" style="left: 0px; position: absolute;">ABI</label><textarea primary="" id="contractAbi" rows="5"></textarea></div>
                  </div>
                </div>
              </div>
              <hr role="separator" aria-orientation="horizontal" class="my-3 v-divider theme--light">
              Or upload a json file containing a <b>contractName</b> field and/or an <b>abi</b> field (a Truffle or Hardhat build file for example):<br> <input type="file" id="fileUploader" accept="application/json">
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text" id="updateContract"><span class="v-btn__content">Update</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;
